pkg/proxy/topology_test.go

```diff
@@ -0,0 +1,198 @@
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package proxy
+
+import (
+	"k8s.io/apimachinery/pkg/types"
+	"reflect"
+	"testing"
+)
+
+func TestFilterTopologyEndpoint(t *testing.T) {
+	testCases := []struct {
+		nodeMap         NodeMap
+		endpoints       []Endpoint
+		currentNodeName types.NodeName
+		topologyKeys    []string
+		expected        []Endpoint
+	}{
+		{
+			// Case[0]: no endpoint
+			nodeMap: NodeMap{
+				"testNode1": &BaseNodeInfo{
+					name: "testNode",
+					labels: map[string]string{
+						"kubernetes.io/hostname":                   "10.0.0.1",
+						"failure-domain.beta.kubernetes.io/zone":   "90001",
+						"failure-domain.beta.kubernetes.io/region": "sg",
+					},
+				},
+			},
+			endpoints:       []Endpoint{},
+			currentNodeName: "testNode",
+			topologyKeys:    []string{"failure-domain.beta.kubernetes.io/region"},
+			expected:        []Endpoint{},
+		},
+		{
+			// Case[1]: no topologyKeys
+			nodeMap: NodeMap{
+				"testNode1": &BaseNodeInfo{
+					name: "testNode1",
+					labels: map[string]string{
+						"kubernetes.io/hostname":                 "10.0.0.1",
+						"failure-domain.beta.kubernetes.io/zone": "90001",
+					},
+				},
+				"testNode2": &BaseNodeInfo{
+					name:   "testNode2",
+					labels: map[string]string{},
+				},
+			},
+			endpoints: []Endpoint{
+				&BaseEndpointInfo{Endpoint: "1.1.1.1:11", NodeName: "testNode1"},
+				&BaseEndpointInfo{Endpoint: "1.1.1.2:11", NodeName: "testNode2"},
+			},
+			currentNodeName: "testNode1",
+			topologyKeys:    []string{},
+			expected: []Endpoint{
+				&BaseEndpointInfo{Endpoint: "1.1.1.1:11", NodeName: "testNode1"},
+				&BaseEndpointInfo{Endpoint: "1.1.1.2:11", NodeName: "testNode2"},
+			},
+		},
+		{
+			// Case[2]: normal topology key with hard requirement
+			nodeMap: NodeMap{
+				"testNode1": &BaseNodeInfo{
+					name: "testNode1",
+					labels: map[string]string{
+						"kubernetes.io/hostname":                 "10.0.0.1",
+						"failure-domain.beta.kubernetes.io/zone": "90001",
+					},
+				},
+				"testNode2": &BaseNodeInfo{
+					name: "testNode2",
+					labels: map[string]string{
+						"kubernetes.io/hostname":                 "10.0.0.2",
+						"failure-domain.beta.kubernetes.io/zone": "90002",
+					},
+				},
+				"testNode3": &BaseNodeInfo{
+					name: "testNode3",
+					labels: map[string]string{
+						"kubernetes.io/hostname":                 "10.0.0.3",
+						"failure-domain.beta.kubernetes.io/zone": "90001",
+					},
+				},
+			},
+			endpoints: []Endpoint{
+				&BaseEndpointInfo{Endpoint: "1.1.1.1:11", NodeName: "testNode1"},
+				&BaseEndpointInfo{Endpoint: "1.1.1.2:11", NodeName: "testNode2"},
+			},
+			currentNodeName: "testNode3",
+			topologyKeys:    []string{"kubernetes.io/hostname", "failure-domain.beta.kubernetes.io/zone"},
```

Comment:
Stepping back, here are the cases I think we need are a bit more, and we can isolate the tests better:

// Prove base case
1) No endpoints at all = 0 endpoints
2) No topology key, 2 nodes each with 2 endpoints = 4 endpoints

// Prove 1 topology key works
3) 1 topology key (hostname), 2 nodes each with 2 endpoints, 1 match = 2 endpoints
4) 1 topology key (hostname), 2 nodes each with 2 endpoints, no match = 0 endpoints
5) 1 topology key (zone), 2 nodes in zone a, 2 nodes in zone b, each with 2 endpoints = 4 endpoints

// Prove 2 topology keys work
6) 2 topology keys (hostname, zone), 2 nodes each with 2 endpoints, 1 hostname match = 2 endpoints // 2nd key ignored
7) 2 topology keys (hostname, zone), 2 nodes in zone a, 2 nodes in zone b, each with 2 endpoints, no hostname match, 1 zone match = 4 endpoints
8) 2 topology keys (hostname, zone), 2 nodes in zone a, 2 nodes in zone b, each with 2 endpoints, no hostname match, no zone match = 0 endpoints

// Prove catchall works
9) repeat 3 with a catchall = 2 endpoints
9) repeat 4 with a catchall = 2 endpoints
10) repeat 8 with a catchall = 4 endpoints

Reply:
done